### PR TITLE
[dhctl] Localhost bootstrap without SSH

### PR DESCRIPTION
**Pull request by mvasl**
## Description

Implemented support for bootstrapping the machine dhctl runs on without using SSH.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: dhctl
type: feature
summary: Support for localhost bootstrapping.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/static'...

**Comment by mvasl** [comment]
/e2e/abort static;Static;containerd;1.27 10472890126 10472890126-1-con-1-27 /sys/deckhouse-oss/install:pr9404 astra@192.168.199.231

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by mvasl** [comment]
/e2e/abort openstack;Standard;containerd;1.27 10472890123 10472890123-1-con-1-27 /sys/deckhouse-oss/install:pr9404 redos@5.159.101.2

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/static'...

**Comment by mvasl** [comment]
/e2e/abort static;Static;containerd;1.27 10473552073 10473552073-1-con-1-27 /sys/deckhouse-oss/install:pr9404 astra@192.168.199.223

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by mvasl** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 10473960482 10473960482-1-con-1-27 /sys/deckhouse-oss/install:pr9404 redos@51.250.94.216

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by mvasl** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 10474643282 10474643282-1-con-1-27 /sys/deckhouse-oss/install:pr9404 redos@89.169.150.101

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by name212** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 10510964996 10510964996-1-con-1-27 /sys/deckhouse-oss/install:pr9404 redos@51.250.4.162

**Comment by name212** [comment]
/e2e/abort static;Static;containerd;1.27 10510964573 10510964573-1-con-1-27 /sys/deckhouse-oss/install:pr9404 astra@192.168.199.250

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by name212** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 10512523170 10512523170-1-con-1-27 /sys/deckhouse-oss/install:pr9404 redos@89.169.147.9

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by name212** [comment]
/e2e/abort static;Static;containerd;1.27 10513781694 10513781694-1-con-1-27 /sys/deckhouse-oss/install:pr9404 astra@192.168.199.105

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/openstack'...

**Comment by mvasl** [comment]
Manual local bootstrap (debug logging was enabled in dhctl)
Снимок экрана 2024-08-27 в 13 46 40

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/gcp'...

**Comment by mvasl** [comment]
Installer config as tested
```
apiVersion: deckhouse.io/v1
kind: ClusterConfiguration
clusterType: Static
podSubnetCIDR: 10.111.0.0/16
serviceSubnetCIDR: 10.222.0.0/16
kubernetesVersion: "Automatic"
clusterDomain: "cluster.local"
---
apiVersion: deckhouse.io/v1
kind: InitConfiguration
deckhouse:
  devBranch: pr9404
  imagesRepo: <dev registry repo>
  registryDockerCfg: <...>
---
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: deckhouse
spec:
  version: 1
  enabled: true
  settings:
    bundle: Default
    logLevel: Info
---
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: global
spec:
  version: 1
  settings:
    modules:
      publicDomainTemplate: "%s.192-168-1-210.nip.io"
---
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: user-authn
spec:
  version: 1
  enabled: true
  settings:
    controlPlaneConfigurator:
      dexCAMode: DoNotNeed
    publishAPI:
      enable: true
      https:
        mode: Global
        global:
          kubeconfigGeneratorMasterCA: ""
---
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: cni-cilium
spec:
  version: 1
  enabled: true
  settings:
    tunnelMode: VXLAN
---
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: flant-integration
spec:
  version: 1
  enabled: false
```

**Comment by mvasl** [comment]
Converge is ok
Снимок экрана 2024-08-27 в 16 01 27
Снимок экрана 2024-08-27 в 16 23 51

File: dhctl/cmd/dhctl/commands/bootstrap/phase.go [hunk 1]
```diff
@@ -36,13 +38,13 @@ func DefineBootstrapInstallDeckhouseCommand(parent *kingpin.CmdClause) *kingpin.
 	app.DefineDeckhouseInstallFlags(cmd)
 
 	cmd.Action(func(c *kingpin.ParseContext) error {
-		var sshClient *ssh.Client
+		var nodeInterface node.Interface = local.NewNodeInterface()
 		if len(app.SSHHosts) != 0 {
```
Comment (name212) [inline dhctl/cmd/dhctl/commands/bootstrap/phase.go:42]: please keep only non local bootstrap for phases

File: dhctl/pkg/config/meta.go [hunk 2]
```diff
@@ -234,6 +234,11 @@ func (m *MetaConfig) FindTerraNodeGroup(nodeGroupName string) []byte {
 	return nil
 }
 
+func (m *MetaConfig) IsStatic() bool {
+	return m.ClusterType == "Static"
+	// return len(m.StaticClusterConfig) > 0 && len(m.ProviderClusterConfig) == 0
```
Comment (name212) [inline dhctl/pkg/config/meta.go:239]: remove comment please

File: dhctl/cmd/dhctl/commands/bootstrap/phase.go (hunk 1 above)
Comment (mvasl) [reply dhctl/cmd/dhctl/commands/bootstrap/phase.go]: done

File: dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go [hunk 3]
```diff
@@ -17,6 +17,8 @@ package bootstrap
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/system/node/local"
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go:20]: not sorted

File: dhctl/pkg/system/node/local/node.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+type NodeInterface struct{}
+
+func NewNodeInterface() *NodeInterface {
+	return &NodeInterface{}
+}
+
+func (n *NodeInterface) Command(name string, args ...string) node.Command {
+	log.DebugLn("Starting NodeInterface.command")
+	log.DebugLn("Stop NodeInterface.command")
```
Comment (name212) [inline dhctl/pkg/system/node/local/node.go:30]: defer

File: dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go (hunk 3 above)
Comment (mvasl) [reply dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go]: fixed

File: dhctl/pkg/system/node/local/node.go (hunk 4 above)
Comment (mvasl) [reply dhctl/pkg/system/node/local/node.go]: done

**Review by name212 — APPROVED** [review]

**Pull request merged**